Blog post about KUDO Cassandra GA (#267)
kudobuilder/www: content/blog/blog-2020-07-10-cassandra-ga.md

Some highlights from this release are:

* Backup & Restore
* Node Replace/Node Recovery
* Multi-Datacenter Cluster

<!-- more -->

## Backup & Restore

For KUDO Cassandra, we have integrated [Cassandra Medusa](https://github.com/thelastpickle/cassandra-medusa) to allow easy backup and restore operations.
Medusa is a backup system for Cassandra that provides support for backups to Amazon S3, Google Cloud Storage and local storage. At the moment we have only integrated the AWS S3 solution, but can easily add support for the other storage providers, supported by [Apache libcloud](https://libcloud.readthedocs.io/en/stable/storage/supported_providers.html)

With version 1.0.0, it is possible to do a full cluster backups and restore the backups to a new instance of the cluster - it is not possible to restore single nodes, but as Cassandra is a distributed database, a full cluster restore is a more probable use case.

To enable backup and restore functionality, there is a single parameter: `BACKUP_RESTORE_ENABLED`. This will deploy a sidecar container on each Cassandra Node which contains Medusa. The actual backup is triggered by a Job that execs Medusa in the sidecar.

> For a restore, the parameter sets up an init container before the actual start of the container which downloads the backup and initialises the data for that node.

Reviewer: ```suggestion
For a restore, the parameter sets up an init container that will run before the actual C* node container, download the backup and initialize the data for that node.
```

The node eviction works similarly, but without a Kubernetes node failure. To evict a node the recovery controller watches for a specific label with the name `kudo-cassandra/evict` on the pods. If this label is found, the handling is the same: The PVC is removed and the pod is rescheduled.

KUDO Cassandra maintains a mapping for each Cassandra node - when the replacement node is scheduled on a new Kubernetes node, the bootstrap process of KUDO Cassandra detects that it is a replacement node an starts Cassandra in replace mode, which triggers the redistribution of the mirrored data in the Cassandra cluster.

More Details about [node eviction](https://github.com/mesosphere/kudo-cassandra-operator/blob/master/docs/evicting-nodes.md) and [failure handling](https://github.com/mesosphere/kudo-cassandra-operator/blob/master/docs/managing.md#failure-handling) can be found in the documentation.

## Multi-Datacenter Cluster

> KUDO Cassandra allows to create Multi-Datacenter clusters. A single parameter, `NODE_TOPOLOGY` can be set, for example like this:

Reviewer: ```suggestion
KUDO Cassandra allows to create Multi-Datacenter clusters. A single parameter, `NODE_TOPOLOGY` can be set e.g.:
```

      rackLabelValue: us-east-1b
```

This will set up a Cassandra cluster with two rings of 12 nodes each. Each datacenter and each rack can have specific labels. The operator will set up a stateful set for each datacenter, create the correct `rack-dc.properties`, adjust the seed node configuration, etc.

Another option to set up a multi-dc cluster is to use `EXTERNAL_SEED_NODES` to connect two KUDO Cassandra clusters. In this case we would have to deploy two or more instances of KUDO Cassandra, and connect them via the parameter. 

The details on how to set up the different versions can be found in the [documentation](https://github.com/mesosphere/kudo-cassandra-operator/blob/master/docs/multidatacenter.md).

## Planned Features

There are a couple of planned features for KUDO Cassandra, although we are not fully sure about the priorities - If you're going to use or evaluate KUDO Cassandra, let us know what you like or need. 

* [Cassandra Reaper](http://cassandra-reaper.io/) for managing repair maintenance 
* Improvements to Backup & Restore for single nodes
* Automatic scale down and decommissioning
* Other improvements waiting for features in KUDO:

#### Toggle Tasks

> [Toggle Tasks](https://kudo.dev/docs/developing-operators/tasks.html#toggle-task) were recently added to KUDO and allow easy feature toggles. It enables us to simplify the operator structure a bit.

Reviewer: ```suggestion
[Toggle Tasks](https://kudo.dev/docs/developing-operators/tasks.html#toggle-task) were recently added to KUDO and allow easy feature toggles. Using it with the C* operator will allow us to simplify the operator structure a bit.
```

---
date: 2020-07-10
---

# KUDO Cassandra 1.0.0 Released

We have released (KUDO Cassandra)[https://github.com/mesosphere/kudo-cassandra-operator] 1.0.0 for general availability. 

Some highlights from this release are:

* Backup & Restore
* Node Replace/Node Recovery
* Multi-Datacenter Cluster

<!-- more -->

## Backup & Restore

For KUDO Cassandra, we have integrated (Cassandra Medusa)[https://github.com/thelastpickle/cassandra-medusa] to allow easy backup and restore operations.

> Medusa is a backup system for Cassandra that provides support for backups to Amazon S3, Google Cloud Storage and local storage. At the moment we have only integrated the AWS S3 solution, but can easily add support for the other storage providers.

Reviewer 2: ```suggestion
Medusa is a backup system for Cassandra that provides support for backups to Amazon S3, Google Cloud Storage and local storage. At the moment we have only integrated the AWS S3 solution, but can easily add support for the other storage providers, supported by [Apache libcloud](https://libcloud.readthedocs.io/en/stable/storage/supported_providers.html)
```

To enable backup and restore functionality, there is a single parameter: `BACKUP_RESTORE_ENABLED`. This will deploy a sidecar container on each Cassandra Node which contains Medusa. The actual backup is triggered by a Job that execs Medusa in the sidecar.

For a restore, the parameter sets up an init container before the actual start of the container which downloads the backup and initialises the data for that node.

Details about this feature can be found (here)[https://github.com/mesosphere/kudo-cassandra-operator/blob/master/docs/backup.md].

## Node Eviction/Node Recovery

One of the tasks of running a bigger Cassandra cluster is handling node failures and node replacement. There are generally two options for the storage provider. Either elastic block storage, which makes it easy to have the same data available on multiple servers, but limits the performance and is an additional complex software to manage. The second option is local storage, which provides the best performance, but has the risk of data loss in case of failure.

> The data loss is not really a problem for Cassandra - as long as the replication factor for a keyspace is setup correctly, copies of the data are stored on different nodes. But local storage requires manual intervention when a Cassandra node really fails. The cluster needs to be informed that a node was lost and was replaced, in case of Kubernetes the persistent volume and persistent volume claims need to be cleaned up, etc.

Reviewer 2: ```suggestion
Apache Cassandra is designed to tolerate node failure - as long as the minimum number of required nodes are up and running and the replication factor for a keyspace is setup correctly, copies of the data are stored on different nodes. But local storage requires manual intervention when a Cassandra node really fails. The cluster needs to be informed that a node was lost and was replaced, in case of Kubernetes the persistent volume and persistent volume claims need to be cleaned up, etc.
```
